source/simulator/entropy_viscosity.cc

```diff
@@ -0,0 +1,439 @@
+    double global_for_sum[2], global_for_max[2];
+
+    dealii::Utilities::MPI::sum (local_for_sum, mpi_communicator, global_for_sum);
+    dealii::Utilities::MPI::max (local_for_max, mpi_communicator, global_for_max);
+
+    const double average_entropy = global_for_sum[0] / global_for_sum[1];
+
+    // return the maximal deviation of the entropy everywhere from the
+    // average value
+    return std::max(global_for_max[1] - average_entropy,
+                    average_entropy - (-global_for_max[0]));
+  }
+
+
+
+  template <int dim>
+  double
+  Simulator<dim>::
+  compute_viscosity (internal::Assembly::Scratch::AdvectionSystem<dim> &scratch,
+                     const double                        global_u_infty,
+                     const double                        global_field_variation,
+                     const double                        average_field,
+                     const double                        global_entropy_variation,
+                     const double                        cell_diameter,
+                     const AdvectionField               &advection_field) const
+  {
+    // discontinuous Galerkin doesn't require an artificial viscosity
+    if (advection_field.is_discontinuous(introspection))
+      return 0.;
+
+    std::vector<double> residual = assemblers->compute_advection_system_residual(*scratch.material_model_inputs.cell,
+                                                                                 advection_field,
+                                                                                 scratch);
+
+    double max_residual = 0;
+    double max_velocity = 0;
+    double max_density = (advection_field.is_temperature()) ? 0.0 : 1.0;
+    double max_specific_heat = (advection_field.is_temperature()) ? 0.0 : 1.0;
+    double max_conductivity = 0;
+
+    for (unsigned int q=0; q < scratch.finite_element_values.n_quadrature_points; ++q)
+      {
+        const Tensor<1,dim> velocity = (scratch.old_velocity_values[q] +
+                                        scratch.old_old_velocity_values[q]) / 2;
+
+        const double strain_rate = ((scratch.old_strain_rates[q]
+                                     + scratch.old_old_strain_rates[q]) / 2).norm();
+
+        if (parameters.stabilization_alpha == 2)
+          {
+            const double field = (scratch.old_field_values[q] + scratch.old_old_field_values[q]) / 2;
+            residual[q] *= std::abs(field - average_field);
+          }
+
+        max_residual = std::max (residual[q],     max_residual);
+        max_velocity = std::max (velocity.norm()
+                                 + parameters.stabilization_gamma * strain_rate * cell_diameter,
+                                 max_velocity);
+
+        if (advection_field.is_temperature())
+          {
+            max_density = std::max       (scratch.material_model_outputs.densities[q],              max_density);
```

Comment:
We already do. `Simulator<dim>::get_artificial_viscosity` replaces the real density by the reference density if we use the reference density profile (entropy_viscosity.cc line 383)